[ES|QL] [Discover] Creating where clause filters from the table, sidebar and table row viewer

stratoula opened the pull request:
## Summary

Part of https://github.com/elastic/kibana/issues/181280

This PR handles the creation of filters (where clause) in Discover ES|QL mode. The parts that are being handled are:

- sidebar
- document viewer
- table

![meow](https://github.com/elastic/kibana/assets/17003240/f2e32e91-5d76-4723-93c1-dbeadb7eb9cb)

The creation of filters from the charts is not here.

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

elasticmachine commented:
Pinging @elastic/kibana-esql (Team:ESQL)

jughosta commented:
> FTR Configs #61 / discover/group4 discover esql view filtering by clicking on the table should append a where clause by clicking the table

Looks like tests are flaky

stratoula commented:
Yes I want to check them Julia 👍

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
Flaky test runner 50 times https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5807

stratoula commented:
/ci

jughosta commented on packages/kbn-esql-utils/src/utils/append_to_query.test.ts:
```diff
@@ -5,21 +5,116 @@
+    });
+
+    it('appends a where clause in an existing query checking that the value is not null if the user asks for existence', () => {
+      expect(appendWhereClauseToESQLQuery('from logstash-* // meow', '_exists_', 'dest', '+')).toBe(
```
If we are using 

```appendWhereClauseToESQLQuery('from logstash-* // meow', 'dest', 'tada!', '+', 'string')``` 
and 
```appendWhereClauseToESQLQuery('from logstash-* // meow', 'dest', 'tada!', '-', 'string')``` 

for +/- actions, would be great to pass params for "exists" filter as:

```
appendWhereClauseToESQLQuery('from logstash-* // meow', 'dest',  undefined, '_exists_')
```

instead of 

```
appendWhereClauseToESQLQuery('from logstash-* // meow', '_exists_', 'dest', '+')

```

to stay consistent with field name and operation arguments.

stratoula replied:
The weirdness start from Discover to be fair but I see your point 😄 I will adapt.

stratoula replied:
https://github.com/elastic/kibana/pull/181399/commits/a6ddb094a5490fe1f2255d6cb4961e0817ace9f5

jughosta commented on packages/kbn-unified-data-table/src/components/__snapshots__/data_table_columns.test.tsx.snap:
```diff
@@ -468,7 +319,7 @@ Array [
     "id": "message",
     "isSortable": true,
-    "schema": "string",
+    "schema": "kibana-json",
```
It seems to be incorrect that schema got changed in the tests. What did trigger it?

stratoula replied:
You were not applying columnsMeta in the test, this is why it has this value. I updated the test to give the correct configuration so now the snapshot will be correct.

stratoula replied:
https://github.com/elastic/kibana/pull/181399/commits/a6ddb094a5490fe1f2255d6cb4961e0817ace9f5

jughosta commented on src/plugins/discover/public/application/main/components/layout/discover_layout.tsx:
```diff
@@ -145,6 +147,27 @@ export function DiscoverLayout({ stateContainer }: DiscoverLayoutProps) {
+          operation,
+          fieldType
+        );
+        data.query.queryString.setQuery({
```
Can we also track clicks as in `onAddFilter`?

stratoula replied:
Sure! I will use a different event `esql_filter_added` unless you want to use the same but I feel that a different event will be easier to track, no?

stratoula replied:
https://github.com/elastic/kibana/pull/181399/commits/a6ddb094a5490fe1f2255d6cb4961e0817ace9f5

jughosta commented on src/plugins/discover/public/application/main/components/layout/discover_layout.tsx:
```diff
@@ -318,7 +346,7 @@ export function DiscoverLayout({ stateContainer }: DiscoverLayoutProps) {
                 onAddField={onAddColumn}
                 columns={currentColumns}
-                onAddFilter={!isPlainRecord ? onAddFilter : undefined}
+                onAddFilter={!isPlainRecord ? onAddFilter : onPopulateWhereClause}
```
Let's create a variable to reuse it here and above for `mainDisplay`

stratoula replied:
https://github.com/elastic/kibana/pull/181399/commits/a6ddb094a5490fe1f2255d6cb4961e0817ace9f5

jughosta commented on src/plugins/discover/public/application/main/components/sidebar/lib/get_field_list.ts:
```diff
@@ -73,7 +73,7 @@ export function getTextBasedQueryFieldList(
         type: column.meta?.type ?? 'unknown',
         esTypes: column.meta?.esType ? [column.meta?.esType] : undefined,
-        searchable: false,
+        searchable: true,
```
I think we check `filterable` to decide whether to show filter buttons. Where do we check `searchable`?

stratoula replied:
If you check the code the filterable is dependent on searchable

```
  public get filterable() {
    return (
      this.name === '_id' ||
      this.scripted ||
      ((this.spec.indexed || this.searchable) && this.kbnFieldType.filterable)
    );
  }
```
we are doing this in other places too and we used this flag to disable the filtering

jughosta commented on src/plugins/discover/public/components/doc_table/doc_table_wrapper.tsx:
```diff
@@ -72,7 +72,7 @@ export interface DocTableProps {
    * Filter callback
    */
-  onFilter: DocViewFilterFn;
+  onFilter?: DocViewFilterFn;
```
What makes this prop optional for the legacy table?

stratoula replied:
It is a mess here, it is optional in other places and required to others. I dont allow the onFilter on the embeddable so this means that it should be optional. This change is fixing this

jughosta commented on test/functional/apps/discover/group6/_sidebar_field_stats.ts:
```diff
@@ -298,6 +341,11 @@ export default function ({ getService, getPageObjects }: FtrProviderContext) {
+        await PageObjects.unifiedFieldList.clickFieldListMinusFilter('enabled', 'true');
+        await testSubjects.click('TextBasedLangEditor-expand');
+        const editorValue = await monacoEditor.getCodeEditorValue();
+        expect(editorValue).to.eql(`row enabled = true\n| where \`enabled\`!=true`);
```
Would be great to not mix field stats tests and the new filters for ES|QL. Can we extract it into a separate test suit (example: `_esql_where_filters`)? 

The new `filtering by clicking on the table` could be also moved out to a such new file.

stratoula replied:
Julia I am so sorry but I dont agree on this. Filtering is part of the stats so I dont see any reason to create a new file for this. We are going to add more effort to the CI and increase the time without any reason. I am trying to create new files only if there is a good reason and this is not I feel. :/

stratoula replied:
Let's start like that and if we have problems in the future let's re-evaluate

jughosta commented on packages/kbn-esql-utils/src/utils/append_to_query.ts:
```diff
@@ -5,9 +5,73 @@
+  // casting to string
+  // there are some field types such as the ip that need
+  // to cast in string first otherwise ES will fail
+  if (fieldType !== 'string' && fieldType !== 'number' && fieldType !== 'boolean') {
+    fieldName = `${fieldName}::string`;
+  }
```
Can we limit this functionality to field types we are sure about (string, number, ip etc) and hide filter buttons for all others? 

<img width="1579" alt="Screenshot 2024-04-30 at 14 38 04" src="https://github.com/elastic/kibana/assets/1415710/e46d26ee-2e60-4a2a-b6f3-306f1480c84b">

jughosta commented on packages/kbn-esql-utils/src/utils/append_to_query.ts:
```diff
@@ -5,9 +5,73 @@
+  // if where command already exists in the end of the query:
+  // - we need to append with and if the filter doesnt't exist
+  // - we need to change the filter operator if the filter exists with different operator
+  // - we do nothing if the filter exists with the same operator
```
It allowed to add `agent` twice. Is it expected?

<img width="1201" alt="Screenshot 2024-04-30 at 14 43 16" src="https://github.com/elastic/kibana/assets/1415710/60f524a6-c8ad-4fa2-986e-c4f85e41fbc5">

stratoula replied:
Yes I didn't want to complicate it more. I dont expect many users to click the existence of the same field multiple times but I am open to discussion if you feel is necessary

stratoula replied:
I actually did that as it was quite straightforward https://github.com/elastic/kibana/pull/181399/commits/193e6708d46b590b23b42073188ab1bbddca1e49

jughosta commented on packages/kbn-unified-data-table/src/components/data_table_columns.tsx:
```diff
@@ -118,7 +118,17 @@ function buildEuiGridColumn({
-  const dataViewField = dataView.getFieldByName(columnName);
+  const dataViewField = !isPlainRecord
+    ? dataView.getFieldByName(columnName)
+    : new DataViewField({
+        name: columnName,
+        type: columnsMeta?.[columnName]?.type ?? 'unknown',
+        esTypes: columnsMeta?.[columnName]?.esType
+          ? ([columnsMeta[columnName].esType] as string[])
+          : undefined,
+        searchable: true,
+        aggregatable: false,
+      });
```
I think we need now a separate util function for creating a DataViewField instance for text-based columns to make sure that UnifiedDataTable and UnifiedFieldList (via src/plugins/discover/public/application/main/components/sidebar/lib/get_field_list.ts) have the same input.

Also we could probably control what field we want to be filterable in a single place.

stratoula replied:
I can do that, but as I explain above I dont want to control it because with casting we can be sure that the filtering will work

stratoula replied:
I think this should be handled by the Discovery team, this way of creating fields for text based languages pre-exist my PR. I don't think it should be handled by this PR.

stratoula replied:
I created an issue to track it https://github.com/elastic/kibana/issues/182646 as it will be better to be handled on a separate PR

jughosta reviewed (commented):
Very nice feature!

drewdaemon commented on packages/kbn-esql-utils/src/utils/append_to_query.ts:
```diff
@@ -5,9 +5,73 @@
+    }
+    // filter does not exist in the where clause
+    const whereClause = `and ${fieldName}${operator}${filterValue}`;
+    return appendToESQLQuery(baseESQLQuery, whereClause);
```
This string manipulation approach looks good for now, but I think it also highlights a current limitation. We have code to go from string to AST, but not from AST to string. 

If we did, we could make edits to the AST directly before serializing it into a new query. This would probably be more robust. If we start seeing more of these tasks, it's something we could explore.

stratoula replied:
Agreed

drewdaemon commented on packages/kbn-esql-utils/src/utils/append_to_query.ts:
```diff
@@ -5,9 +5,73 @@
+    const whereCommand = ast[ast.length - 1];
+    const whereAstText = whereCommand.text;
+    // the filter already exists in the where clause
+    if (whereAstText.includes(field) && whereAstText.includes(String(filterValue))) {
+      const pipesArray = baseESQLQuery.split('|');
+      const whereClause = pipesArray[pipesArray.length - 1];
+      const matches = whereClause.match(new RegExp(field + '(.*)' + String(filterValue)));
```
Instead of re-parsing the query, I think we should rely on the AST here.

The current code breaks in a few scenarios

## False positive: detecting an existing filter

For example, the check at 54 is `true` for this scenario.
```
        appendWhereClauseToESQLQuery(
          `from kibana_sample_data_logs 
        | limit 10 
        | where CIDR_MATCH(ip1, "127.0.0.2/32", "127.0.0.3/32")`,
          'ip',
          '127.0.0.2/32',
          '-',
          'string'
        )
```

It then applies the regular expression and adds the operator in place of the first comma, resulting in an invalid query
```
from kibana_sample_data_logs·
            | limit 10·
            | where CIDR_MATCH(ip!= \"127.0.0.2/32\", \"127.0.0.3/32\")
```

## Replaces a nested boolean expression

In the following example, a boolean expression is successfully detected, but it isn't at the top level of the `WHERE` clause so it shouldn't be changed.
```
appendWhereClauseToESQLQuery(
          `from kibana_sample_data_logs | limit 10
          | where case(agent.keyword == "foo", 1, 0) == 1`,
          'agent.keyword',
          'foo',
          '-',
          'string'
        )
```

Results in

```
from kibana_sample_data_logs | limit 10
              | where case(agent.keyword != \"foo\", 1, 0) == 1
```

If we instead inspect the AST and find a compatible operator which has the `field` and `value` as its arguments, it will be a more robust existence check and you can know at what level the comparison is occurring (i.e. that it isn't nested somewhere).

Unfortunately, I guess we'll still have to rely on the final string manipulation until we have code that takes the AST back to a string.

stratoula replied:
@drewdaemon this was a path I tried to take in the beginning but I decided against it as this needs to be a recursive function which thanx to our types mess is very difficult to create. It sounds to me a very complicated and possibly imperfomant way to find what you want while you can do it easier with a regex.

About your first example, this was a bug I fixed here and also added a test https://github.com/elastic/kibana/pull/181399/commits/a6ddb094a5490fe1f2255d6cb4961e0817ace9f5

About your second one I think is ok and also you cant easily detect it even with ast.

To prove you this you can check this commit https://github.com/elastic/kibana/pull/181399/commits/8c143964924051480ae366d37733fddbedfaf1b3. You can. see the complexity of the recursive function and our types mess. I would personally prefer to go with the previous implementation but I am happy to hear your thoughts on that.

drewdaemon replied:
If you're okay with the second scenario not being covered, I think it's fair to stick with your current implementation for now. 

What I think this really highlights is the need for a good AST traversal API so that you don't have to worry about writing a complicated routine and can instead use existing code e.g. 

```ts
const nodes = findAll(whereNode, 'function', ['==', '!=']);
const matchingTopLevelClause = nodes.find(node => node.arguments[0].name == field && node.arguments[1].name == value && node.depth == 1);

...
```

Or maybe

```ts
const matcher = (node) => node.depth == 1
  && isFunctionNode(node) 
  && ['==', '!='].includes(node.name)
  && node.arguments[0].name == field
  && node.arguments[1].name == value;

const existingClause = findInAst(whereNode, matcher);

...
```

stratoula replied:
I absolutely agree ++

drewdaemon reviewed (commented):
Overall working nicely, but I think there are some corner cases to look at

stratoula commented:
/ci

drewdaemon reviewed (approved):
ES|QL stuff looks good to me. I created a couple issues to improve our query manipulation DX
- https://github.com/elastic/kibana/issues/182255
- https://github.com/elastic/kibana/issues/182257

drewdaemon commented:
Lol... I guess I technically approved the whole thing since I'm a member of both Discover and ES|QL teams.... 

I only meant my review to be on behalf of the ES|QL team. I leave the Discover stuff to @jughosta

stratoula commented:
Thanx Drew, I want to revert and i will def not merge without Julia's approval :D

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

jughosta reviewed (commented):
The functionality looks great!

One thing before the approval: can we update the following tests? Currently they generate incorrect schemas and might be less effective in testing:
`packages/kbn-unified-data-table/src/components/data_table_columns.test.tsx`

stratoula commented on packages/kbn-unified-data-table/src/components/default_cell_actions.tsx:
```diff
@@ -20,22 +20,23 @@ function onFilterCell(
 
-  if (field && context.onFilter) {
-    context.onFilter(field, value, mode);
+  const filterField = context.isPlainRecord ? field : context.dataView.fields.getByName(columnId);
+
+  if (filterField && context.onFilter) {
+    context.onFilter(filterField, value, mode);
```
I dont get the suggestion Julia, can you elaborate please?

stratoula replied:
Done https://github.com/elastic/kibana/pull/181399/commits/fff036df37c55946dab4a0577c977e8dc2b8a5a6

jughosta replied:
From what I can see we are passing `field` from an upper context. Is not it correct there?

stratoula replied:
Yeah makes sense, your comment confused me because you are using `field` in the check but `filterField` in the function 😄

stratoula replied:
Done https://github.com/elastic/kibana/pull/181399/commits/adc9d6eb8141685a23d9d9d35f65d092dc91ee4e

stratoula commented on packages/kbn-unified-data-table/src/components/__snapshots__/data_table_columns.test.tsx.snap:
```diff
@@ -3180,7 +2800,7 @@ Array [
     "initialWidth": 212,
     "isSortable": true,
-    "schema": "datetime",
+    "schema": "kibana-json",
```
Done https://github.com/elastic/kibana/pull/181399/commits/fff036df37c55946dab4a0577c977e8dc2b8a5a6

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
Ok so Julia on your comments:

- simplified the code https://github.com/elastic/kibana/pull/181399/commits/adc9d6eb8141685a23d9d9d35f65d092dc91ee4e
- fixed the tests that were not already correct in main https://github.com/elastic/kibana/pull/181399/commits/fff036df37c55946dab4a0577c977e8dc2b8a5a6 (we were also marking extension as number which seems wrong to me, I changed it to string)
- created an issue about the utility you suggested https://github.com/elastic/kibana/issues/182646. I think it needs a discussion first and should not be part of this PR
- about the FT I get your way of thinking. I suggest to move on with one FT and if we feel it creates problems in the future we can split.

stratoula commented:
/ci

jughosta reviewed (approved):
Awesome work! And thanks for addressing the comments! 🚢

stratoula commented:
/ci
